### PR TITLE
Get rid of StatusService.

### LINKED ISSUE #7198: Letting phone go to sleep while any non-home screen is open crashes Status when relaunched
### User Story

As a user, I want to be able to keep using the app after it goes to sleep and wakes.

### Description

*Type*: Bug

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: 
When an Android device running Status goes to sleep with a dapp or chat open, reopening it crashes the app.

screenshot_20181229-120732

#### Expected behavior
[comment]: # (Describe what you expected to happen.)

The app does not crash.

#### Actual behavior
[comment]: # (Describe what actually happened.)

The app crashes.

### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status on Android
- Open any dapp
- Sleep your phone (either via timeout or just press the power button)
- If you wake it within 5 or so minutes, everything is fine.
- If you wake it after 5 or so minutes, it crashes from limbo (a state when an app is neither killed off due to non-usage, nor has active cache due to several minutes of inactivity).

### Solution
[comment]: # (Please summarise the solution and provide a task list on what needs to be fixed.)
*Summary*: N/A

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: 0.9.33 (2018122003)
[comment]: # (Android or iOS.)
* Operating System: Android

### LINKED ISSUE #7281: Random 'java.lang.NullPointerException' at attempt to start 0.9.32-1
### Description

*Type*: Bug
*Summary*: sometimes at attempt to start build 0.9.32-1 'java.lang.NullPointerException: Attempt to invoke virtual method 'void android.os.Messenger.send(android.os.Message)' on a null object reference'. Can't reproduce manually, can be random.

#### Expected behavior
can start app

#### Actual behavior
sdfsadf

### Reproduction

- Open Status

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: release 0.9.32-1 (staging fleet is set by default, used this [custom build](https://ci.status.im/job/status-react/job/release/job/release%252F0.9.32-1/lastSuccessfulBuild/artifact/pkg/StatusIm-181217-093744-07a7a5-e2e.apk) 
 Status version: release 0.9.32-1 (beta fleet is set by default)
* Test in SauceLab (staging): https://app.saucelabs.com/tests/0b8d26472cf5459d8d3becac3a036aa3/commands?auth=05a66084424f252494761160912f052d#22
* Operating System: Android

#### Logs

[logcat-1.txt](https://github.com/status-im/status-react/files/2760329/logcat-1.txt) (staging)
[logcat.txt](https://github.com/status-im/status-react/files/2782430/logcat.txt) (beta)

### PR DESCRIPTION
**Pull request by mandrigin**
Android 9 somehow changed the restrictions of starting background services in `Activity#onResume`, it isn't allowed anymore. I'm not exactly sure if that is intended or a bug.

UPD: it's an Android bug: 
* https://issuetracker.google.com/issues/113122354
* https://issuetracker.google.com/issues/110237673
* https://issuetracker.google.com/issues/117790600

~~This change delays starting the service up until the app is in foreground, by using lifecycle listeners (see [this article](https://developer.android.com/reference/android/arch/lifecycle/Lifecycle) to read more on them).~~

~~That is what I used as an example of this workaround: https://github.com/mapbox/mapbox-events-android/pull/157~~

Okay, StatusService was only used as a weird layer between status-go to pass `signalEvent:`. I simplified it in a way that is still compatible with old status-go (no class rename), but it works completely differently.

In a nutshell, we just use a singleton class that either gets initialised when the `StatusModule` is there or when JNI is looking up for an instance. Then it just passes all the `signalEvent:` calls to `StatusModule`.

That way we don't need for service to initialise, so I removed all mentions of it and Android works roughly the same as iOS now.

fixes #7198 
fixes #7281
 
status: ready

### CODE REVIEW

**Review by rasom — DISMISSED** [review]

File: modules/react-native-status/android/build.gradle [hunk 1]
```diff
@@ -6,11 +6,11 @@ def getStatusGoVersion = { ->
 }
 
 android {
-    compileSdkVersion 26
+    compileSdkVersion 28
 
     defaultConfig {
-        minSdkVersion 18
-        targetSdkVersion 26
+        minSdkVersion 23
```
Comment (pedropombeiro) [inline modules/react-native-status/android/build.gradle:12]: Does this mean that the lowest Android version we support is 6.0?

**Review by pedropombeiro — DISMISSED** [review]

File: modules/react-native-status/android/build.gradle (hunk 1 above)
Comment (mandrigin) [reply modules/react-native-status/android/build.gradle:12]: The app itself already requires [`minSdkVersion 23`](https://github.com/status-im/status-react/blob/develop/android/app/build.gradle#L141), so to make `react-native-status` do that, it's no problem.

**Comment by mandrigin** [comment]
@Swader hey! since you were along the people who were able to reproduce this issue quite easily, would you mind checking if this PR actually solves it?

Keep in mind, that that will most probably overwrite your current version of the app. It is based on our nightlies, so it should be pretty solid, but you can always lose data with it... So, do it only if you have your account seed phrase backed up and keep in mind that you might lose your contacts there...

**Comment by Swader** [comment]
@mandrigin I would love to but need help on how to do this as I don't know how to build and test a PR version of Status for Android. Any guide I can reference?

**Comment by mandrigin** [comment]
@Swader if you are okay with all the precautions, then you can just download and install this APK (the latest one for android on the [table of builds](
https://github.com/status-im/status-react/pull/7334#issuecomment-457195142): 
https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190124-140803-6e4d3c-pr.apk

But again, I want to reiterate that:
> Keep in mind, that that will most probably overwrite your current version of the app. It is based on our nightlies, so it should be pretty solid, but you can always lose data with it... So, do it only if you have your account seed phrase backed up and keep in mind that you might lose your contacts there...

Or, if you are using nightlies, maybe you can just wait for the next nightly and this fix will be included in it.


<blockquote></blockquote>

**Comment by mandrigin** [comment]
Unfortunately, right now it is impossible to install PR builds aside the release, but this PR will fix it: https://github.com/status-im/status-react/pull/7287

**Comment by Swader** [comment]
Android just gives me "app not installed" after downloading and clicking Install and confirming an update process. No other error or explanation.

**Comment by mandrigin** [comment]
Oh, thanks to Android for not being helpful.
Are you currently using a nightly or a google play release?

**Comment by Swader** [comment]
I am currently on last night's nightly, installed without a problem just an hour ago.

**Comment by mandrigin** [comment]
@Swader any chance you have `adb` installed on your computer?

**Comment by Swader** [comment]
I do now. What do I need to do?

**Comment by mandrigin** [comment]
@Swader 

1. download the file to your computer:
> https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190124-140803-6e4d3c-pr.apk

2. open terminal/console

3. connect your phone with a cable to the computer, make sure that `adb list` shows it

4. to to the dir with this file

5. do `adb install StatusIm-190124-140803-6e4d3c-pr.apk`


<blockquote></blockquote>

**Comment by Swader** [comment]
`Failure [INSTALL_FAILED_VERSION_DOWNGRADE]`

**Comment by mandrigin** [comment]
@Swader ah, ~~not~~ now it's more useful :-) It just claims that your nightly (from today) is more fresh than this builds (from yesterday). Let me rebuild it for you, then it should be fine.

**Comment by mandrigin** [comment]
@Swader try installing this one: https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190125-114343-6e4d3c-pr.apk

<blockquote></blockquote>

**Comment by Swader** [comment]
No effect. Or should I say same effect, different message.

**Comment by mandrigin** [comment]
@Swader thats... interesting. Can you tell me the exact steps you used to reproduce it and the model/maker of your Android device as well as if the firmware installed (stock or some open-source installed).

Also, `adb logcat` output would be useful too.

**Comment by mandrigin** [comment]
`adb logcat`'s output should be acquired like that: https://sites.google.com/site/androidhowto/how-to-1/save-logcat-to-a-text-file

`adb logcat -d > logcat.txt`

<blockquote><div><strong><a href="https://sites.google.com/site/androidhowto/how-to-1/save-logcat-to-a-text-file">Save LogCat To A Text File - Android How-to's</a></strong></div><div>Resources and How-to's for Google's Android SDK</div></blockquote>

**Comment by mandrigin** [comment]
it really looks like that this is an Android 9 bug, many people are having this issue: https://issuetracker.google.com/issues/113122354

<blockquote><div><strong><a href="https://issuetracker.google.com/issues/113122354">Sign in - Google Accounts</a></strong></div></blockquote>

**Comment by mandrigin** [comment]
and that one: https://issuetracker.google.com/issues/110237673

<blockquote><div><strong><a href="https://issuetracker.google.com/issues/110237673">Sign in - Google Accounts</a></strong></div></blockquote>

**Comment by mandrigin** [comment]
and that one: https://issuetracker.google.com/issues/117790600

<blockquote><div><strong><a href="https://issuetracker.google.com/issues/117790600">Sign in - Google Accounts</a></strong></div></blockquote>

**Comment by mandrigin** [comment]
marking the issue as WIP again, time to take a look at AOSP

**Comment by mandrigin** [comment]
@annadanchenko @rasom @PombeirP I ended up getting rid of `.StatuService` whatsoever. It obviously changes everything here, so a re-review and re-test is needed :-)

**Comment by mandrigin** [comment]
it should also fix https://github.com/status-im/status-react/issues/7281

**Comment by mandrigin** [comment]
@Swader can you try this build and check if you can reproduce the issue? 

https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190125-144936-59485e-pr.apk

<blockquote></blockquote>

**Comment by Swader** [comment]
OnePlus 6T, stock. 

Reproduce: Go anywhere in the app (in this case localethereum) that isn't the home screen and then let the phone go to sleep. Wake 30 seconds later and it will always appear. Relevant: https://github.com/status-im/status-react/issues/7198

Logcat:
[logcat3.zip](https://github.com/status-im/status-react/files/2796721/logcat3.zip)

Testing new build in a few mins.

**Comment by Swader** [comment]
Problem resolved, thank you :)
I'll try and reproduce but right now, not succeeding. Well done!

**Comment by mandrigin** [comment]
@Swader awesome!


Btw, looking on your logcat. I don't know what Google did with Android 9.0 but services are really broken:

1. Exception from **Gitter.im**: `java.lang.RuntimeException: Unable to resume activity {im.gitter.gitter/im.gitter.gitter.activities.MainActivity}: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=im.gitter.gitter/.content.RestService (has extras) }: app is in background uid UidRecord{45aaa01 u0a136 TPSL bg:+23m25s654ms idle change:idle procs:1 seq(0,0,0)}` 

2. Exception from something from **Netgear**: `01-21 11:30:57.578 31694 31694 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {com.netgear.android/com.netgear.android.activity.MainScreenActivity}: java.lang.IllegalStateException: Not allowed to start service Intent { act=com.netgear.android.AudioStreamService.ACTION_BACKGROUND cmp=com.netgear.android/.stream.AudioStreamService (has extras) }: app is in background uid UidRecord{35e788e u0a117 TPSL bg:+10m10s437ms idle change:idle procs:1 seq(0,0,0)}`

3. Us (Status.im) `01-22 18:55:01.301 14242 14242 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {im.status.ethereum/im.status.ethereum.MainActivity}: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=im.status.ethereum/.module.StatusService }: app is in background uid UidRecord{cca0891 u0a124 TPSL bg:+8m55s99ms idle change:cached procs:1 seq(0,0,0)}`

**Comment by mandrigin** [comment]
@Swader thanks a lot for testing the app!

**Comment by mandrigin** [comment]
history rewrite to sign the commit and get autotests to run :)

**Review by rasom — APPROVED** [review]
12m

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusService.java [hunk 2]
```diff
@@ -9,98 +9,55 @@
...
-    }
-
-    private static class IncomingHandler extends Handler {
-
-        private final WeakReference<StatusService> service;
-
-        IncomingHandler(StatusService service) {
+/**
+ * StatusService has nothing to do with Android services anymore.
+ * The name "StatusService" is kep to keep backward compatibility with status-go.
```
Comment (pedropombeiro) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusService.java]:
```suggestion
 * The name "StatusService" is kept to keep backward compatibility with status-go.
```

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 3]
```diff
@@ -366,6 +340,7 @@ public void moveToInternalStorage(Callback callback) {
     public void startNode(final String config) {
         Log.d(TAG, "startNode");
         if (!checkAvailability()) {
+            Log.e(TAG, "[startNode] Activity doesn't exists, cannot start node");
```
Comment (pedropombeiro) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java]:
```suggestion
            Log.e(TAG, "[startNode] Activity doesn't exist, cannot start node");
```

**Review by pedropombeiro — APPROVED** [review]
Great cleanup! Only a couple minor ortho fixes

**Comment by annadanchenko** [comment]
Update: can reproduce on nightly, so not introduced by this PR

@mandrigin sometimes there is an error or crash upon logout (investigating exact case when it happens and trying to reproduce also on nightly), account has public, group and 1:1 chats and performed some transactions. Pre-selected mailserver is used (mail-03.ac-cn-hongkong-c.eth.beta), develop mode is off.

these steps can lead to the issue:

preselect mailserver and logout/login again
1.  join some public chat with messages
2. log out
3. log in (so it's not automatic)
4. open chat you've just joined
5. as soon as you see green Connected  line go to profile and logout

```
01-28 12:03:19.204 17622 18304 D StatusModule: stopNode
01-28 12:03:19.206 17622 17663 D StatusService: [signalEvent] event: {"type":"discovery.stopped","event":null}
01-28 12:03:19.206 17622 17663 D StatusService: [signalEvent] passing event to the listener: {"type":"discovery.stopped","event":null}
01-28 12:03:19.206 17622 17663 D StatusModule: [handleEvent] event: {"type":"discovery.stopped","event":null}
01-28 12:03:19.206 17622 17669 D ReactNativeJS: DEBUG [status-im.utils.handlers:34] - Handling re-frame event:  :accounts.login.callback/get-user-password-success
01-28 12:03:19.211 17622 17669 D ReactNativeJS: DEBUG [status-im.utils.handlers:34] - Handling re-frame event:  :signals/signal-received
01-28 12:03:19.213 17622 17669 D ReactNativeJS: DEBUG [status-im.events:808] - :event-str {"type":"discovery.stopped","event":null}
01-28 12:03:19.219 17622 17669 D ReactNativeJS: DEBUG [status-im.utils.handlers:34] - Handling re-frame event:  :keychain.callback/can-save-user-password?-success
01-28 12:03:19.221 17622 18195 D StatusService: [signalEvent] event: {"type":"node.stopped","event":null}
01-28 12:03:19.221 17622 18195 D StatusService: [signalEvent] passing event to the listener: {"type":"node.stopped","event":null}
01-28 12:03:19.221 17622 18195 D StatusModule: [handleEvent] event: {"type":"node.stopped","event":null}
01-28 12:03:19.224 17622 17669 I ReactNativeJS: PRETTY PRINTED EXCEPTION
01-28 12:03:19.224 17622 17669 I ReactNativeJS: 
01-28 12:03:19.224 17622 17669 I ReactNativeJS: ***
01-28 12:03:19.224 17622 17669 I ReactNativeJS: NAME: "Error"
01-28 12:03:19.224 17622 17669 I ReactNativeJS: MESSAGE: fx/defn expects a map of cofx as first argument got {:event [:accounts.login.callback/get-user-password-success nil], :db nil, :now 1548673399206} in function function (e) {switch(arguments.length){case 1:return status_im.accounts.login.core.open_login_callback.cljs$core$IFn$_invoke$arity$1(arguments[0]);case 2:return status_im.accounts.login.core.open_login_callback.cljs$core$IFn$_invoke$arity$2(arguments[0],arguments[1]);default:throw Error(["Invalid arity: ",cljs.core.str.cljs$core$IFn$_invoke$arity$1(arguments.length)].join(""))}}
01-28 12:03:19.224 17622 17669 I ReactNativeJS: 
01-28 12:03:19.224 17622 17669 I ReactNativeJS: cljs$core$IFn$_invoke$arity$2@index.android.bundle:12:6630778
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:12:7886655
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:12:1739573
01-28 12:03:19.224 17622 17669 I ReactNativeJS: invoke_interceptor_fn@index.android.bundle:12:1669879
01-28 12:03:19.224 17622 17669 I ReactNativeJS: invoke_interceptors@index.android.bundle:12:1670180
01-28 12:03:19.224 17622 17669 I ReactNativeJS: execute@index.android.bundle:12:1672629
01-28 12:03:19.224 17622 17669 I ReactNativeJS: handle@index.android.bundle:12:1681042
01-28 12:03:19.224 17622 17669 I ReactNativeJS: re_frame$router$IEventQueue$_process_1st_event_in_queue$arity$1@index.android.bundle:12:1708127
01-28 12:03:19.224 17622 17669 I ReactNativeJS: re_frame$router$IEventQueue$_run_queue$arity$1@index.android.bundle:12:1693112
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:12:1696292
01-28 12:03:19.224 17622 17669 I ReactNativeJS: re_frame$router$IEventQueue$_fsm_trigger$arity$3@index.android.bundle:12:1700503
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:12:1694414
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:24:1801
01-28 12:03:19.224 17622 17669 I ReactNativeJS: k@index.android.bundle:24:643
01-28 12:03:19.224 17622 17669 I ReactNativeJS: q@index.android.bundle:24:1007
01-28 12:03:19.224 17622 17669 I ReactNativeJS: callImmediates@index.android.bundle:24:3202
01-28 12:03:19.224 17622 17669 I ReactNativeJS: value@index.android.bundle:19:2904
01-28 12:03:19.224 17622 17669 I ReactNativeJS: index.android.bundle:19:1377
01-28 12:03:19.224 17622 17669 I ReactNativeJS: value@index.android.bundle:19:2621
01-28 12:03:19.224 17622 17669 I ReactNativeJS: value@index.android.bundle:19:1347
01-28 12:03:19.224 17622 17669 I ReactNativeJS: [native code]
01-28 12:03:19.224 17622 17669 I ReactNativeJS: value@index.android.bundle:19:1282
01-28 12:03:19.224 17622 17669 I ReactNativeJS: [native code]
01-28 12:03:19.224 17622 17669 I ReactNativeJS: 
01-28 12:03:19.224 17622 17669 I ReactNativeJS: ***
01-28 12:03:19.226 17622 17669 I ReactNativeJS: PRETTY PRINTED EXCEPTION

```

[Status.log](https://github.com/status-im/status-react/files/2802728/Status.log)
[geth.log](https://github.com/status-im/status-react/files/2802730/geth.log)

Another set of smaller logs after issue is reproduced:
[logcat.log](https://github.com/status-im/status-react/files/2802749/logcat.log)
[geth.log](https://github.com/status-im/status-react/files/2802752/geth.log)
[Status.log](https://github.com/status-im/status-react/files/2802754/Status.log)
